test/integration/apimachinery/admission_test.go

```diff
@@ -0,0 +1,871 @@
+	}
+	if name != h.expectName {
+		return
+	}
+
+	h.recorded[phase] = request
+}
+
+func (h *holder) verify(t *testing.T) {
+	h.lock.Lock()
+	defer h.lock.Unlock()
+
+	if err := h.verifyRequest(h.recorded[mutation]); err != nil {
+		t.Errorf("mutation error: %v", err)
+	}
+	if err := h.verifyRequest(h.recorded[validation]); err != nil {
+		t.Errorf("validation error: %v", err)
+	}
+}
+
+func (h *holder) verifyRequest(request *v1beta1.AdmissionRequest) error {
+	if request == nil {
+		return fmt.Errorf("no request received")
+	}
+
+	if h.expectObject {
+		if err := h.verifyObject(request.Object.Object); err != nil {
+			return fmt.Errorf("object error: %v", err)
+		}
+	} else if request.Object.Object != nil {
+		return fmt.Errorf("unexpected object: %#v", request.Object.Object)
+	}
+
+	if h.expectOldObject {
+		if err := h.verifyObject(request.OldObject.Object); err != nil {
+			return fmt.Errorf("old object error: %v", err)
+		}
+	} else if request.OldObject.Object != nil {
+		return fmt.Errorf("unexpected old object: %#v", request.OldObject.Object)
+	}
+
+	return nil
+}
+
+func (h *holder) verifyObject(obj runtime.Object) error {
+	if obj == nil {
+		return fmt.Errorf("no object sent")
+	}
+	if obj.GetObjectKind().GroupVersionKind() != h.expectGVK {
+		return fmt.Errorf("expected %#v, got %#v", h.expectGVK, obj.GetObjectKind().GroupVersionKind())
+	}
+	return nil
+}
+
+// TestAdmission tests communication between API server and webhook process.
+func TestWebhookV1beta1(t *testing.T) {
+	// holder communicates expectations to webhooks, and results from webhooks
+	holder := &holder{t: t}
+
+	// set up webhook server
+	roots := x509.NewCertPool()
+	if !roots.AppendCertsFromPEM(localhostCert) {
+		t.Fatal("Failed to append Cert from PEM")
+	}
+	cert, err := tls.X509KeyPair(localhostCert, localhostKey)
+	if err != nil {
+		t.Fatalf("Failed to build cert with error: %+v", err)
+	}
+	webhook := &http.Server{
+		Addr: ":22443",
+		TLSConfig: &tls.Config{
+			RootCAs:      roots,
+			Certificates: []tls.Certificate{cert},
+		},
+	}
+	handler := func(phase string) {
+		http.HandleFunc("/"+phase, func(w http.ResponseWriter, r *http.Request) {
+			var body []byte
+			if r.Body != nil {
+				if data, err := ioutil.ReadAll(r.Body); err == nil {
+					body = data
+				}
+			}
+			r.Body.Close()
+			contentType := r.Header.Get("Content-Type")
+			if contentType != "application/json" {
+				t.Errorf("contentType=%s, expect application/json", contentType)
+				return
+			}
+			// The AdmissionReview that was sent to the webhook
+			requestedAdmissionReview := v1beta1.AdmissionReview{}
+
+			// The AdmissionReview that will be returned
+			responseAdmissionReview := v1beta1.AdmissionReview{}
+			if err := json.Unmarshal(body, &requestedAdmissionReview); err != nil {
+				t.Errorf("Fail to deserialize object: %s with error: %+v", string(body), err)
+				responseAdmissionReview.Response = response(false, err.Error())
+			} else {
```

Comment:
Can we capture line 328-335 in a function, then call the function here and return early to reduce the nesting?